packages/flutter/lib/src/widgets/sliver.dart

```diff
@@ -987,14 +1115,32 @@ class SliverMultiBoxAdaptorElement extends RenderObjectElement implements Render
           _childElements.remove(index);
         }
       }
-      // processElement may modify the Map - need to do a .toList() here.
-      _childElements.keys.toList().forEach(processElement);
+
+      for (int index in _childElements.keys.toList()) {
+        final Key key = _childElements[index].widget.key;
+        final int newIndex = key == null ? -1 : widget.delegate.findIndexByKey(key);
+        if (newIndex != -1 && newIndex != index) {
+          newChildren[newIndex] = _childElements[index];
+          // We need to make sure the original index gets processed.
+          newChildren.putIfAbsent(index, () => null);
+          // We do not want the remapped child to get deactivated during processElement.
+          _childElements.remove(index);
+        } else {
+          newChildren.putIfAbsent(index, () => _childElements[index]);
+        }
+      }
+
+      renderObject.debugChildIntegrityEnabled = false; // Moving children will temporary violate the integrity.
+      newChildren.keys.forEach(processElement);
       if (_didUnderflow) {
         final int lastKey = _childElements.lastKey() ?? -1;
+        // There should be no swapping.
```

Comment:
deal with under flow case

Reply:
hat do you mean by "there should be no swapping"?